File: therapist/src/types/notification.ts
/**
 * 通知相关类型定义
 */

export enum NotificationType {
  NEW_ORDER = 'new_order',
  ORDER_CANCELLED = 'order_cancelled',
  ORDER_COMPLETED = 'order_completed',
  SYSTEM_MESSAGE = 'system_message',
  PAYMENT_SUCCESS = 'payment_success',
}

export enum NotificationStatus {
  PENDING = 'pending',
  SENT = 'sent',
  FAILED = 'failed',
  READ = 'read',
}

export enum NotificationPriority {
  LOW = 'low',
  NORMAL = 'normal',
  HIGH = 'high',
  URGENT = 'urgent',
}

export interface Notification {
  id: number;
  type: NotificationType;
  priority: NotificationPriority;
  title: string;
  body: string;
  data: Record<string, any>;
  status: NotificationStatus;
  sent_via: string | null;
  sent_at: string | null;
  read_at: string | null;
  created_at: string;
}

export interface NotificationListResponse {
  notifications: Notification[];
  total: number;
  unread_count: number;
  page: number;
  page_size: number;
}

export interface NotificationSettings {
  notifications_enabled: boolean;
  sound_enabled: boolean;
  vibration_enabled: boolean;
  new_order_enabled: boolean;
  order_cancelled_enabled: boolean;
  order_completed_enabled: boolean;
  system_message_enabled: boolean;
  new_order_sound: string | null;
  new_order_vibration_pattern: string | null;
  do_not_disturb_periods: Record<string, any> | null;
}

export interface UpdateNotificationSettingsRequest {
  notifications_enabled?: boolean;
  sound_enabled?: boolean;
  vibration_enabled?: boolean;
  new_order_enabled?: boolean;
  order_cancelled_enabled?: boolean;
  order_completed_enabled?: boolean;
  system_message_enabled?: boolean;
  new_order_sound?: string;
  new_order_vibration_pattern?: string;
  do_not_disturb_periods?: Record<string, any>;
}

export interface PushTokenRequest {
  token: string;
  device_id?: string;
  device_name?: string;
  platform: 'ios' | 'android' | 'web';
  app_version?: string;
}

// WebSocket 消息类型
export interface WebSocketMessage {
  type: 'connected' | 'notification' | 'pong' | 'error';
  message?: string;
  notification?: {
    type: string;
    title: string;
    body: string;
    data: Record<string, any>;
    priority: string;
    timestamp: string;
  };
  therapist_id?: number;
  timestamp?: string;
}

File: backend/app/models/notification.py
"""
通知模型
"""
from datetime import datetime
from typing import Optional
from sqlalchemy import String, Integer, Boolean, DateTime, Text, JSON, Enum as SQLEnum, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship
import enum

from app.core.database import Base


class NotificationType(str, enum.Enum):
    """通知类型"""
    NEW_ORDER = "new_order"              # 新订单
    ORDER_CANCELLED = "order_cancelled"  # 订单取消
    ORDER_COMPLETED = "order_completed"  # 订单完成
    SYSTEM_MESSAGE = "system_message"    # 系统消息
    PAYMENT_SUCCESS = "payment_success"  # 支付成功


class NotificationStatus(str, enum.Enum):
    """通知状态"""
    PENDING = "pending"  # 待发送
    SENT = "sent"        # 已发送
    FAILED = "failed"    # 发送失败
    READ = "read"        # 已读


class NotificationPriority(str, enum.Enum):
    """通知优先级"""
    LOW = "low"       # 低优先级
    NORMAL = "normal" # 普通优先级
    HIGH = "high"     # 高优先级
    URGENT = "urgent" # 紧急


class PushToken(Base):
    """技师推送 Token"""
    __tablename__ = "push_tokens"
    
    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    therapist_id: Mapped[int] = mapped_column(ForeignKey("therapists.id"), unique=True, index=True)
    expo_push_token: Mapped[str] = mapped_column(String(200), index=True)
    device_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    device_name: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    platform: Mapped[str] = mapped_column(String(20))  # ios/android/web
    app_version: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


class Notification(Base):
    """通知记录"""
    __tablename__ = "notifications"
    
    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    therapist_id: Mapped[int] = mapped_column(ForeignKey("therapists.id"), index=True)
    type: Mapped[NotificationType] = mapped_column(SQLEnum(NotificationType), index=True)
    priority: Mapped[NotificationPriority] = mapped_column(
        SQLEnum(NotificationPriority),
        default=NotificationPriority.NORMAL
    )
    
    # 通知内容
    title: Mapped[str] = mapped_column(String(100))
    body: Mapped[str] = mapped_column(Text)
    data: Mapped[dict] = mapped_column(JSON, default={})  # 额外数据（订单ID、跳转路径等）
    
    # 发送状态
    status: Mapped[NotificationStatus] = mapped_column(
        SQLEnum(NotificationStatus),
        default=NotificationStatus.PENDING,
        index=True
    )
    sent_via: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)  # push/websocket/both
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    
    # 时间戳
    sent_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True, index=True)
    read_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, index=True)


class TherapistNotificationSettings(Base):
    """技师通知设置"""
    __tablename__ = "therapist_notification_settings"
    
    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    therapist_id: Mapped[int] = mapped_column(ForeignKey("therapists.id"), unique=True, index=True)
    
    # 全局开关
    notifications_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    sound_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    vibration_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    
    # 分类通知开关
    new_order_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    order_cancelled_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    order_completed_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    system_message_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    
    # 新订单特殊设置
    new_order_sound: Mapped[Optional[str]] = mapped_column(String(50), nullable=True, default="default")
    new_order_vibration_pattern: Mapped[Optional[str]] = mapped_column(String(100), nullable=True, default="default")
    
    # 免打扰时段（JSON 格式）
    # 例如：[{"start": "22:00", "end": "08:00", "days": [0,1,2,3,4,5,6]}]
    do_not_disturb_periods: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)
    
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


# 避免循环导入
from app.models.therapist import Therapist

